src/sql/src/plan/statement/ddl.rs

```diff
@@ -4941,12 +4942,61 @@ pub fn plan_alter_object_rename(
         (ObjectType::ClusterReplica, UnresolvedObjectName::ClusterReplica(name)) => {
             plan_alter_cluster_replica_rename(scx, object_type, name, to_item_name, if_exists)
         }
+        (ObjectType::Schema, UnresolvedObjectName::Schema(name)) => {
+            plan_alter_schema_rename(scx, name, to_item_name, if_exists)
+        }
         (object_type, name) => {
             unreachable!("parser set the wrong object type '{object_type:?}' for name {name:?}")
         }
     }
 }
 
+pub fn plan_alter_schema_rename(
+    scx: &mut StatementContext,
+    name: UnresolvedSchemaName,
+    to_schema_name: Ident,
+    if_exists: bool,
+) -> Result<Plan, PlanError> {
+    let Some((db_spec, schema_spec)) = resolve_schema(scx, name.clone(), if_exists)? else {
+        let object_type = ObjectType::Schema;
+        scx.catalog.add_notice(PlanNotice::ObjectDoesNotExist {
+            name: name.to_ast_string(),
+            object_type,
+        });
+        return Ok(Plan::AlterNoop(AlterNoopPlan { object_type }));
+    };
+
+    // TODO(parkmycar): Make this a structured error and re-evaluate if it needs to be an error.
+    let ResolvedDatabaseSpecifier::Id(db_id) = &db_spec else {
+        sql_bail!("Cannot rename a schema in the ambient database");
+    };
+
+    // Make sure the name is unique.
+    let schemas = scx.catalog.get_database(db_id).schema_ids();
+    let new_schema_name = normalize::ident(to_schema_name);
+    if schemas.contains_key(&new_schema_name) {
+        return Err(PlanError::ObjectAlreadyExists {
+            name: new_schema_name,
+            object_type: ObjectType::Schema,
+        });
+    }
+
+    // Prevent users from renaming system related schemas.
+    let schema = scx.catalog.get_schema(&db_spec, &schema_spec);
+    if schema.owner_id() == MZ_SYSTEM_ROLE_ID {
+        return Err(PlanError::NeverSupported {
+            feature: format!("renaming the {} schema", schema.name().schema),
+            documentation_link: None,
+            details: None,
+        });
```

Comment:
The `bail_never_supported!` macro might be handy here. It's subjective though, so up to you.

Reply:
Good idea, updated!